### PR TITLE
fix: optimize git clone process

### PR DESCRIPTION
**Pull request by hyy0322**
**What this PR does / why we need it**:

- clone with depth flag first, if fail (some version of bitbucket not support), retry without depth flag.

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @zhujian7 @supereagle 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: build/resolver/git/entrypoint.sh [hunk 1]
```diff
@@ -144,7 +141,22 @@ wrapPull() {
         failed=$(mkdir $PULLING_LOCK > /dev/null 2>&1 || echo fail)
         if [[ $failed != "fail" ]]; then
             echo "Got the lock, start to pulling..."
-            pull
+            if ! pull; then
```
Comment (zhujian7) [inline build/resolver/git/entrypoint.sh:144]: This error judgment should be fine-tuned, only *bitbucket* server git clone with a *certain* depth-caused error will be retried.

**Comment by hyy0322** [comment]
Refactor as what we talked about earlier, PTAL again. @zhujian7

File: build/resolver/git/entrypoint.sh [hunk 2]
```diff
@@ -144,7 +142,22 @@ wrapPull() {
         failed=$(mkdir $PULLING_LOCK > /dev/null 2>&1 || echo fail)
         if [[ $failed != "fail" ]]; then
             echo "Got the lock, start to pulling..."
-            pull
+            pull || res=$?
+            if [ $res -eq 99 ]; then
+                echo "Fail to pull with depth flag, retry without depth flag..."
+                unset GIT_DEPTH_OPTION
+                unset GIT_DEPTH_OPTION_DEEPER
+                count=3
```
Comment (zhujian7) [inline build/resolver/git/entrypoint.sh:150]: I think retry once is enough.

File: build/resolver/git/entrypoint.sh [hunk 3]
```diff
@@ -200,18 +213,30 @@ pull() {
...
         # give it a 'oauth2:' prefix to get 'oauth2:<encoded_token>'.
         if [ ! -z ${SCM_AUTH+x} ]; then
-            SCM_URL=$( modifyURL $SCM_URL $SCM_AUTH )
+            SCM_URL_MODIFIED=$( modifyURL $SCM_URL $SCM_AUTH )
         fi
 
         if [[ "${SOURCE_BRANCH}" == "${TARGET_BRANCH}" ]]; then
             echo "Clone $SOURCE_BRANCH..."
-            git clone -v -b master ${GIT_DEPTH_OPTION:-} --single-branch --recursive ${SCM_URL} data
+            git clone -v -b master ${GIT_DEPTH_OPTION:-} --single-branch --recursive ${SCM_URL_MODIFIED} data 2>&1 | tee clone_result.log
```
Comment (zhujian7) [inline build/resolver/git/entrypoint.sh:221]: Please delete the "clone_result.log" file before exit

File: build/resolver/git/entrypoint.sh [hunk 4]
```diff
@@ -200,18 +213,30 @@ pull() {
...
         if [ ! -z ${SCM_AUTH+x} ]; then
-            SCM_URL=$( modifyURL $SCM_URL $SCM_AUTH )
+            SCM_URL_MODIFIED=$( modifyURL $SCM_URL $SCM_AUTH )
         fi
 
         if [[ "${SOURCE_BRANCH}" == "${TARGET_BRANCH}" ]]; then
             echo "Clone $SOURCE_BRANCH..."
-            git clone -v -b master ${GIT_DEPTH_OPTION:-} --single-branch --recursive ${SCM_URL} data
+            git clone -v -b master ${GIT_DEPTH_OPTION:-} --single-branch --recursive ${SCM_URL_MODIFIED} data 2>&1 | tee clone_result.log 
+            if grep "The requested URL returned error: 500" clone_result.log > /dev/null 2>&1; then
```
Comment (zhujian7) [inline build/resolver/git/entrypoint.sh:222]: Make this code block `if ..... fi` to a common func?

File: build/resolver/git/entrypoint.sh [hunk 5]
```diff
@@ -200,18 +213,30 @@ pull() {
...
+                # use code 99 for git clone --depth fail
+                return 99
```
Comment (zhujian7) [inline build/resolver/git/entrypoint.sh:224]: The `99` `98` return codes are unreadable, any other `best practice` we can refer to?

File: build/resolver/git/entrypoint.sh (hunk 4 above)
Comment (hyy0322) [reply build/resolver/git/entrypoint.sh]: done

File: build/resolver/git/entrypoint.sh (hunk 2 above)
Comment (hyy0322) [reply build/resolver/git/entrypoint.sh]: done

File: build/resolver/git/entrypoint.sh (hunk 5 above)
Comment (hyy0322) [reply build/resolver/git/entrypoint.sh]: Shell scripts use int as return code. I think it is as I have already detailed code meaning in comments.

File: build/resolver/git/entrypoint.sh [hunk 6]
```diff
@@ -157,7 +170,7 @@ wrapPull() {
     # Write commit id to output file, which will be collected by Cyclone
     cd $WORKDIR/data
     echo "Collect commit id to result file /__result__ ..."
-    echo "LastCommitID:`git log -n 1 --pretty=format:"%H"`" >> /__result__;
+    echo "LastCommitID:`git log -n 1 --pretty=format:"%H"`" > /__result__;
```
Comment (zhujian7) [inline build/resolver/git/entrypoint.sh:173]: Why do we need to change this?

**Comment by zhujian7** [comment]
And can we make this change only take effect for BitBucket, maybe some code like:
```
if [ "${SCM_TYPE}" != "Bitbucket" ]; then
```

File: build/resolver/git/entrypoint.sh [hunk 7]
```diff
@@ -144,7 +142,22 @@ wrapPull() {
         failed=$(mkdir $PULLING_LOCK > /dev/null 2>&1 || echo fail)
         if [[ $failed != "fail" ]]; then
             echo "Got the lock, start to pulling..."
-            pull
+            pull || res=$?
+            if [ $res -eq 99 ]; then
+                echo "Fail to pull with depth flag, retry without depth flag..."
+                unset GIT_DEPTH_OPTION
+                unset GIT_DEPTH_OPTION_DEEPER
+                count=2
```
[the remark this answers is not in the document]
Comment (hyy0322) [reply build/resolver/git/entrypoint.sh:150]: conut=2 means pull with GIT_DEPTH_OPTION_DEEPER twice. It means pull with GIT_DEPTH_OPTION_DEEPER retry one time.

File: build/resolver/git/entrypoint.sh (hunk 6 above)
Comment (hyy0322) [reply build/resolver/git/entrypoint.sh:173]: Don't need to append commit info to __result__. Just write to it is enough.

**Comment by hyy0322** [comment]
> And can we make this change only take effect for BitBucket, maybe some code like:
> 
> ```
> if [ "${SCM_TYPE}" != "Bitbucket" ]; then
> ```

I think it's ok. Retry logic can fit all SCM. Though it only take effect on bitbucket. If we code like
```if [ "${SCM_TYPE}" != "Bitbucket" ]; then```,
there may be redundant code like
```
if [ "${SCM_TYPE}" != "Bitbucket" ]; then
  pull
  retry
else
  pull
fi
```

**Pull request merged**